File: demo/tasks/base_task.py
from typing import List, NamedTuple, Optional, Callable
import abc
import asyncio
import logging

from expansion import types
from expansion import modules


class JournalRecord(NamedTuple):
    timestamp: int
    # When this record has been added to the journal (ingame milliseconds time)
    action: str
    # What is happening or what has happened

    def __str__(self):
        return f"{self.timestamp}: {self.action}"


class BaseTask:
    global_id: int = 0

    def __init__(self, name: str,
                 system_clock: modules.SystemClock,
                 write_logs: bool = True):
        BaseTask.global_id += 1
        self.id = BaseTask.global_id
        self.name = name
        self.journal: List[JournalRecord] = []
        self.system_clock: modules.SystemClock = system_clock
        self.write_logs = write_logs
        self.finished: bool = False
        self._asyncio_task: Optional[asyncio.Task] = None
        self._logger = logging.getLogger(f"{name}:{self.id}")
        self._complete_cb: Optional[Callable[[bool], None]] = None

    def add_journal_record(self, action: str):
        record = JournalRecord(self.system_clock.cached_time(), action)
        self.journal.append(record)
        if self.write_logs:
            self._logger.info(record)

    async def run(self, *argc, **argv) -> bool:
        self.finished = False
        self.add_journal_record("Started")
        status: bool = False
        try:
            status = await self._impl(*argc, **argv)
            self.add_journal_record("Finished" if status else "Failed")
        except asyncio.CancelledError:
            self.add_journal_record("Canceled")
        except asyncio.TimeoutError:
            self.add_journal_record("Canceled by timeout")
        self.finished = True
        if self._complete_cb:
            self._complete_cb(status)
        return status

    def run_async(self,
                  complete_cb: Optional[Callable[[bool], None]] = None,
                  *argc, **argv):
        self._complete_cb = complete_cb
        self._asyncio_task = asyncio.create_task(self.run(*argc, **argv))

    async def join(self, timeout: Optional[float] = None):
        if self._asyncio_task:
            await asyncio.wait_for(self._asyncio_task, timeout=timeout)

    @abc.abstractmethod
    async def _impl(self, *argc, **argv) -> bool:
        pass

    def interrupt(self):
        if self._asyncio_task is None:
            return
        if self._asyncio_task.cancel():
            self.add_journal_record("Interrupted")
        else:
            self.add_journal_record("Interrupting failed: task is not running")

File: typescript-sdk/demo/harvester/tasks/base_task.ts
import type { SystemClock } from "../../../highlevel/index.js";
import { create_logger } from "../log.js";
import { is_abort_error } from "../util.js";

export type JournalRecord = {
    timestamp: bigint;
    action: string;
};

export abstract class BaseTask {
    static global_id = 0;

    readonly id: number;
    readonly name: string;
    readonly journal: JournalRecord[] = [];
    finished = false;

    protected readonly system_clock: SystemClock;
    protected readonly write_logs: boolean;
    private readonly _logger;
    private _abort?: AbortController;
    private _running?: Promise<boolean>;
    private _complete_cb?: (status: boolean) => void;

    constructor(name: string, system_clock: SystemClock, write_logs = true) {
        BaseTask.global_id += 1;
        this.id = BaseTask.global_id;
        this.name = name;
        this.system_clock = system_clock;
        this.write_logs = write_logs;
        this._logger = create_logger(`${name}:${this.id}`);
    }

    protected get signal(): AbortSignal | undefined {
        return this._abort?.signal;
    }

    add_journal_record(action: string): void {
        const record: JournalRecord = {
            timestamp: this.system_clock.now_us(),
            action,
        };
        this.journal.push(record);
        if (this.write_logs) {
            this._logger.info(`${record.timestamp}: ${record.action}`);
        }
    }

    async run(): Promise<boolean> {
        this.finished = false;
        this._abort = new AbortController();
        this.add_journal_record("Started");
        let status = false;
        try {
            status = await this._impl();
            this.add_journal_record(status ? "Finished" : "Failed");
        } catch (error) {
            if (is_abort_error(error) || this._abort.signal.aborted) {
                this.add_journal_record("Canceled");
            } else {
                throw error;
            }
        }
        this.finished = true;
        this._complete_cb?.(status);
        return status;
    }

    run_async(complete_cb?: (status: boolean) => void): void {
        this._complete_cb = complete_cb;
        this._running = this.run();
    }

    async join(): Promise<boolean | undefined> {
        return this._running;
    }

    interrupt(): void {
        if (this._abort === undefined) {
            return;
        }
        if (!this._abort.signal.aborted) {
            this.add_journal_record("Interrupted");
            this._abort.abort();
        }
    }

    protected abstract _impl(): Promise<boolean>;
}
